### ISSUE TITLE
KEDA CLI Commands

### ISSUE THREAD

**Issue by jeffhollan** [kedacore/keda#173]
Related closely to #167 I think, but would be nice for the insights that the dashboard provides through some CLI (maybe the `func` CLI, not sure if KEDA needs its own). 

It's possible that the dashboard won't surface much more than the `kubectl` would, but just had an ask from some users that it was hard to know after deploying if things were working or not.  Ideally an experience a bit more than just `kubectl logs` existed.

**Comment by yaron2** [comment]
If I understand correctly, the ask is "how do I monitor KEDA and know if something went wrong with scaling my deployment".

Either cli or ui, currently we have two sources of truth: the controller logs and the status fields of ScaledObjects.

I can assume for the cli a command like ```keda get all``` would show the names of scaled objects, the trigger, and the latest status.

We could also parse the logs of the controller but that is not advisable.
Instead, we should make sure every ScaledObject is updated with human readable info regarding its state.

wdyt?

**Comment by yaron2** [comment]
Relates to #51

**Comment by yaron2** [comment]
Actually,

This is very similar to how you would get visibility into a service mesh.

In that case, I think the correct way to go would be to expose a Restful API from the controllrr to get detailed view of all the scaled objects.

The UI will visualize that data.

A KEDA cli would mean Azure Functions/kubeless/openFaaS/other users would need to have two clis in place. One for their framework and one for KEDA.

If we put that functionality in the func cli, we miss out on all non azure function users.

**Comment by ahmelsayed** [comment]
What kind of information are you thinking about surfacing from keda directly? 

We could have a simple [kubectl plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) and get `kubectl keda ....` style commands? It could wrap kubectl or communicate with that API depending on what data we're interested in showing. Most of the state we have is already stored in the crd for each scaled object.

**Comment by yaron2** [comment]
Yep a kubectl plugin is a good idea - it would communicate with our API that feeds the dashboard as well as direct user queries for more advanced cases.

**Comment by tomkerkhove** [comment]
v1.0 has shipped (:tada:) and I'd love to get working on this for v1.x so we can build the dashboard on top of this.

Anybody who agrees? If so, I can spec it a bit out and provide a proposal.

Given everything gets a dedicated repo, I'd say this should be moved to http://github.com/kedacore/cli once we've agreed on it.

**Comment by tomkerkhove** [comment]
Bumping

**Comment by arschles** [comment]
As I mentioned in #1661, I think there's another possible use case for a `keda` CLI or `kubectl` plugin, and that is to provide users a more convenient way to create `ScaledObject` / `ScaledJob`s (I added a use cases section over in #1661 for those interested). So a CLI could potentially do mutation operations as well.

What do folks think? I can work on this

**Comment by coderanger** [comment]
I don't think this would be a good user experience to encourage. The use of existing tools like Kubectl's generators is highly correlated with new user confusion about how to maintain things over time. They are great for quick demos but terrible for long-term UX :-(

**Comment by zroubalik** [comment]
I tend to agree with @coderanger. 

And there are other aspects for example from maintenance POV: let' say  we implement this kind of CLI, then it should definitely support ScaledObject creation -> thus it should be able to create a Trigger definition and *validate* it on the client side. And because the triggers specs are not some CRDs, but just key value map, you will need to implement the validation checks for all scalers (that's already done in each scaler code) and then track all the changes/additions. It is of course doable (or refactor the existing validation part in the scaler and expose it), but not sure if it is worth the effort. 🤷‍♂️
